dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/FixedSizeClassPathRepository.java

```diff
@@ -0,0 +1,116 @@
+ * caching frequently-used instances.
+ *
+ * <p>The default maximum size is 1000 entries. As per experiments with spring-cloud-gcp project,
+ * this maximum size gives the best performance when running {@link
+ * LinkageChecker#findSymbolProblems()}.
+ *
+ * @see <a href="https://github.com/google/guava/wiki/CachesExplained#size-based-eviction">Guava
+ *     CachesExplained: Size-based Eviction</a>
+ * @see <a href="https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/653"
+ *     >Performance evaluation of FixedSizeClassPathRepository with different maximum cache size</a>
+ */
+final class FixedSizeClassPathRepository extends ClassPathRepository {
+
+  private final Cache<String, JavaClass> loadedClass;
+
+  /**
+   * Mapping from class names to special class file locations.
+   *
+   * <p>While class name and class file name are the same in most cases, sometimes classes are not
+   * placed in the root of a JAR file to support a framework-specific JAR structure. For example,
+   * Spring Boot Gradle Java plugin places class files under "BOOT-INF/classes". To load such
+   * classes by class name, this mapping keeps track of the special location once they are loaded.
+   *
+   * <ul>
+   *   <li>Key: class name (value from {@link JavaClass#getClassName()}) which has a special class
+   *       file name different from its class name. Example: {@code com.google.Foo}
+   *   <li>Value: the special class file name as in {@link JavaClass#getFileName()}, a path that
+   *       locates a class file in a class path. Example: {@code BOOT-INF.classes.com.google.Foo}
+   * </ul>
+   *
+   * @see <a
+   *     href="https://docs.spring.io/spring-boot/docs/current/reference/html/executable-jar.html"
+   *     >Spring Boot Reference Guide: The Executable Jar Format</a>
+   */
+  private final Map<String, String> specialClassFileName;
+
+  FixedSizeClassPathRepository(ClassPath path) {
+    this(path, 1000);
+  }
+
+  @VisibleForTesting
+  FixedSizeClassPathRepository(ClassPath path, long maximumSize) {
+    super(path);
+    loadedClass = CacheBuilder.newBuilder().maximumSize(maximumSize).build();
+    this.specialClassFileName = new HashMap<>();
+  }
+
+  @Override
+  public void storeClass(JavaClass javaClass) {
+    String className = javaClass.getClassName();
+    loadedClass.put(className, javaClass);
+    javaClass.setRepository(this);
+
+    String fileName = javaClass.getFileName();
+    if (!className.equals(fileName)) {
+      // When class file has special location not matching class name, remember it to load the class
+      // file by class name.
+      specialClassFileName.put(className, fileName);
+    }
+  }
+
+  @Override
+  @Nullable
+  public JavaClass findClass(String className) {
+    return loadedClass.getIfPresent(className);
+  }
+
+  @Override
+  public JavaClass loadClass(String className) throws ClassNotFoundException {
+    // Check special location for the class; if none, using className to lookup JavaClass.

```

Comment:
Check special location for the class. If it's not found, lookup by className instead.